### PR TITLE
session driver + refactor

### PR DESCRIPTION
**Pull request by luceos**
0.4 proposal;

- add redis session handler
- allow different db's per service
- more fluent set up

pending fix;

- [x] redis session handler to allow wakeup, sleep

### CODE REVIEW

**Comment by imorland** [comment]
Here's the issues I've encountered testing this locally:

Using this config in `extend.php`

```
$redisConfig = [
    'host' => \getenv('REDIS_HOST'),
    'password' => null,
    'port' => \getenv('REDIS_PORT'),
    'database' => 1,
    'prefix' => 'flarum_'
];

return [
    (new Blomstra\Redis\Extend\Redis($redisConfig))
        ->useDatabaseWith('cache', 1)
        ->useDatabaseWith('queue', 2)
        ->useDatabaseWith('session', 3)
];
```

Or this config:

```
(new Blomstra\Redis\Extend\Redis($redisConfig))->disable(['session'])
```

Both of these setups throw this error:

```
Call to undefined method Blomstra\Redis\Configuration::extend()
thrown in /var/www/vendor/flarum/core/src/Foundation/InstalledSite.php on line 146

Error: Call to undefined method Blomstra\Redis\Configuration::extend() in /var/www/vendor/flarum/core/src/Foundation/InstalledSite.php:146
Stack trace:
#0 [internal function]: Flarum\Foundation\InstalledSite->Flarum\Foundation\{closure}(Object(Flarum\Foundation\Application))
#1 /var/www/vendor/flarum/core/src/Foundation/Application.php(329): call_user_func(Object(Closure), Object(Flarum\Foundation\Application))
#2 /var/www/vendor/flarum/core/src/Foundation/Application.php(270): Flarum\Foundation\Application->fireAppCallbacks(Array)
#3 /var/www/vendor/flarum/core/src/Foundation/InstalledSite.php(150): Flarum\Foundation\Application->boot()
#4 /var/www/vendor/flarum/core/src/Foundation/InstalledSite.php(82): Flarum\Foundation\InstalledSite->bootLaravel()
#5 /var/www/vendor/flarum/core/src/Http/Server.php(58): Flarum\Foundation\InstalledSite->bootApp()
#6 /var/www/vendor/flarum/core/src/Http/Server.php(35): Flarum\Http\Server->safelyBootAndGetHandler()
#7 /var/www/public/index.php(22): Flarum\Http\Server->listen()
#8 {main}
```

File: README.md [hunk 1]
```diff
@@ -1,16 +1,97 @@
...
+            ],
+        ],
+    ]))
+];
+```
+
+
 #### Configuring cache
 
 In your `extend.php`:
```
Comment (imorland) [inline README.md:97]: This can now be removed

**Review by imorland — APPROVED** [review]
Checks out fine with me locally. Confirmed redis content via phpredisadmin.

Only suggestions are:
- Remove old partial instructions from README
- Strongly suggest in README a different redis db for each service, rather than defaulting to the same

**Comment by luceos** [comment]
@imorland i cannot reproduce the cache clearing issue, but I reinstated some older logic; can you try again?

**Comment by imorland** [comment]
@luceos confirmed, the cache issue I experiened yesterday is gone with your latest commit.

Re-approved ;)

**Pull request merged**